packages/wallet/src/redux/protocols/existing-channel-funding/reducer.ts

```diff
@@ -0,0 +1,349 @@
+  if (!checkResult.isSuccess) {
+    return {
+      protocolState: states.failure({ reason: 'Received Invalid Commitment' }),
+      sharedData,
+    };
+  }
+  newSharedData = checkResult.store;
+  if (helpers.isFirstPlayer(protocolState.ledgerId, newSharedData)) {
+    const appPostFundSetupSharedData = craftAndSendAppPostFundCommitment(
+      newSharedData,
+      protocolState.channelId,
+      protocolState.processId,
+    );
+    if (
+      appPostFundSetupSharedData === 'CouldNotSign' ||
+      appPostFundSetupSharedData === 'NotASetupCommitment'
+    ) {
+      return {
+        protocolState: states.failure({ reason: appPostFundSetupSharedData }),
+        sharedData,
+      };
+    } else {
+      newSharedData = appPostFundSetupSharedData;
+    }
+  } else {
+    const ourCommitment = acceptConsensus(theirCommitment);
+    const signResult = signAndStore(newSharedData, ourCommitment);
+    if (!signResult.isSuccess) {
+      return {
+        protocolState: states.failure({ reason: 'Received Invalid Commitment' }),
+        sharedData,
+      };
+    }
+    newSharedData = signResult.store;
+
+    const messageRelay = sendCommitmentReceived(
+      theirAddress(ledgerChannel),
+      processId,
+      signResult.signedCommitment.commitment,
+      signResult.signedCommitment.signature,
+      EXISTING_CHANNEL_FUNDING_PROTOCOL_LOCATOR,
+    );
+    newSharedData = queueMessage(newSharedData, messageRelay);
+  }
+
+  return { protocolState: states.waitForPostFundSetup(protocolState), sharedData: newSharedData };
+};
+
+function craftNewAllocationAndDestination(
+  latestCommitment: Commitment,
+  proposedAmount: string,
+  channelId: string,
+): { proposedAllocation: string[]; proposedDestination: string[] } {
+  const numParticipants = latestCommitment.channel.participants.length;
+  const amountRequiredFromEachParticipant = bigNumberify(proposedAmount).div(numParticipants);
+
+  const proposedAllocation: string[] = [];
+  const proposedDestination: string[] = [];
+
+  for (let i = 0; i < latestCommitment.allocation.length; i++) {
+    const allocation = latestCommitment.allocation[i];
+
+    const newAmount = bigNumberify(allocation).sub(amountRequiredFromEachParticipant);
+
+    if (newAmount.gt('0x0')) {
+      proposedAllocation.push(newAmount.toHexString());
+      proposedDestination.push(latestCommitment.destination[i]);
+    }
+  }
+
+  proposedAllocation.push(proposedAmount);
+  proposedDestination.push(channelId);
+
+  return { proposedAllocation, proposedDestination };
+}
+
+function ledgerChannelNeedsTopUp(latestCommitment: Commitment, proposedAmount: string) {
+  if (latestCommitment.commitmentType !== CommitmentType.App) {
+    throw new Error('Ledger channel is already closed.');
+  }
+  const numParticipants = latestCommitment.channel.participants.length;
+  const amountRequiredFromEachParticipant = bigNumberify(proposedAmount).div(numParticipants);
```

Comment:
I don't understand this logic -- what if I already have enough funds in the channel but you don't? Then this logic says that I have to foot half of your bill?

Reply:
This check is determine if either player needs to top up their ledger channel. If either player needs a top-up both players start the `ledger-topup` protocol which figures out who needs to add more funds and exchanges `commitment`s as necessary.

So in the case of `[A:4,B:6]` and proposedAmount of `10`.  This would return true for both players since `A <10/2`. However in the `ledger-topup` protocol only A would be required to add more funds.

Reply:
But it assumes that each player wants to have the same amount of funds staked in the channel. What if we have a one-way relationship, where I pay you, and I want to top up my end of the channel?

Reply:
At that's right. I've created #545 to address this.